pkg/kubelet/kubelet.go

```diff
@@ -660,9 +650,19 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
 	default:
 		return nil, fmt.Errorf("unsupported CRI runtime: %q", containerRuntime)
 	}
-	runtimeService, imageService, err := getRuntimeAndImageServices(remoteRuntimeEndpoint, remoteImageEndpoint, kubeCfg.RuntimeRequestTimeout)
-	if err != nil {
-		return nil, err
+
+	runtimeService, imageService := kubeDeps.RemoteRuntimeService, kubeDeps.RemoteImageService
+	if runtimeService == nil {
+		runtimeService, err = remote.NewRemoteRuntimeService(remoteRuntimeEndpoint, kubeCfg.RuntimeRequestTimeout.Duration)
```

Comment:
It is a little strange to me to have NewMainKubelet take a `remoteRuntimeEndpoint string` and have RemoteRuntimeService be part of kubeDeps.  Same for image as well.  It could produce confusing behavior for someone trying to override the endpoint if the dependency is set.

Is it possible to initialize the runtimeService and imageService before calling NewMainKubelet, and remove the remote{Runtime,Image}Endpoint arguments?

If it ends up requiring significant refactoring, we can leave it for a follow-up if needed.

Reply:
Seems we can init runtimeService and imageService by endpoint args in this [function](https://github.com/kubernetes/kubernetes/blob/92a14f4f7ef11c580961f0aebe491d0ff148e191/cmd/kubelet/app/server.go#L477) without think too much.
Do you think it's OK?

Reply:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L643 these should be refactor, too

Reply:
> It could produce confusing behavior for someone trying to override the endpoint if the dependency is set.

kubelet code won't init it in dependency now, maybe we should also make a defensive to avoid what you said.

Reply:
> Seems we can init runtimeService and imageService by endpoint args in this function or somewhere else without thinking too much.
Do you think it's OK?

I think it makes sense.

> https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L643 these should be refactored, too

Yeah - that seems a bit more tricky.